### PR TITLE
Allow retrieval of specific type of linked descriptor

### PR DESCRIPTION
**Pull request by WestleyK**
This PR adds a function for retrieval of a specific type of linked descriptor.

### CODE REVIEW

**Comment by dctrud** [comment]
I'm not sure that it's appropriate to filter here. I would expect `GetFromLinkedDescr` to return information from *any* type of linked descriptor, and not filter.

**Comment by dctrud** [comment]
Looking further I believe this will actually break the encryption handling... as that is getting a non DataSignature from `GetFromLinkedDescr` to retrieve the Encrypted key

```
	descr, _, err := img.GetFromLinkedDescr(primDescr.ID)
	if err != nil {
		return nil, errors.Wrapf(err, "retrieving linked descriptors for primary system partition from %s", fn)
	}

	for _, d := range descr {
		if d.Datatype != sif.DataGeneric {
			continue
		}

		// TODO(mem): add a specific data type for encrypted
		// keys. For now, assume the first generic descriptor
		// linking to the primary system partition contains the
		// key
```
https://github.com/sylabs/singularity/blob/dbf3ba164f21fff64343dece730a0d5538c9360b/pkg/util/crypt/key.go#L209

**Comment by WestleyK** [comment]
Good catch @dctrud! I didnt know this function is being used elsewhere :+1:

File: pkg/sif/lookup.go [hunk 1]
```diff
@@ -152,6 +152,31 @@ func (fimg *FileImage) GetSignFromGroup(groupid uint32) ([]*Descriptor, []int, e
 	return descrs, indexes, nil
 }
 
+// GetTypeFromLinkedDescr searches for descriptors that point to "id", only returns the specified type
+func (fimg *FileImage) GetTypeFromLinkedDescr(ID uint32, dataType Datatype) ([]*Descriptor, []int, error) {
```
Comment (mem) [inline pkg/sif/lookup.go]: this function is not returning a Type, it's returning a list of descriptors, filtering them by the ID of the descriptor they link to and their datatype.

*Maybe* something like `GetLinkingDescriptorsByType`? (better suggestions welcomed)

If you want to follow the existing `GetFromLinkedDescr`, maybe `GetFromLinkedDescrByType`?

File: pkg/sif/lookup.go [hunk 2]
```diff
@@ -152,6 +152,31 @@ func (fimg *FileImage) GetSignFromGroup(groupid uint32) ([]*Descriptor, []int, e
...
+	var descrs []*Descriptor
+	var indexes []int
+	var count int
+
+	for i, v := range fimg.DescrArr {
+		if !v.Used {
+			continue
+		} else if v.Datatype == dataType {
```
Comment (mem) [inline pkg/sif/lookup.go]: in general Go prefers less indentation, so:

```
if !v.Used {
    continue
}

if v.Datatype == datatype && v.Link == ID {
    // ...
}
```

File: pkg/sif/lookup.go [hunk 3]
```diff
@@ -152,6 +152,31 @@ func (fimg *FileImage) GetSignFromGroup(groupid uint32) ([]*Descriptor, []int, e
...
+			if v.Link == ID {
+				indexes = append(indexes, i)
+				descrs = append(descrs, &fimg.DescrArr[i])
+				count++
```
Comment (mem) [inline pkg/sif/lookup.go]: note that count is equivalent with `len(descrs)`

File: pkg/sif/lookup.go [hunk 4]
```diff
@@ -152,6 +152,31 @@ func (fimg *FileImage) GetSignFromGroup(groupid uint32) ([]*Descriptor, []int, e
...
+			}
+		}
+	}
+
+	if count == 0 {
```
Comment (mem) [inline pkg/sif/lookup.go]: so here you can say:

```
if len(descrs) == 0 {
    return nil, nil, ErrNotFound
}
```

**Comment by mem** [comment]
sorry, it seems I had this open for a while :-(

**Comment by WestleyK** [comment]
@mem, I'll make a followup PR with your suggestions

**Pull request merged**